src/backend/distributed/connection/remote_commands.c

```diff
@@ -75,6 +75,90 @@ ForgetResults(MultiConnection *connection)
 
 
 /*
+ * NonblockingForgetResults clears a connection from pending activity if doing
+ * so does not require network IO. Returns true if successful, false
+ * otherwise.
+ */
+bool
+NonblockingForgetResults(MultiConnection *connection)
+{
+	PGconn *pgConn = connection->pgConn;
+	bool wasNonblocking = PQisnonblocking(pgConn);
+
+	if (PQstatus(pgConn) != CONNECTION_OK)
+	{
+		return false;
+	}
+
+	while (true)
+	{
+		PGresult *result = NULL;
+
+		CHECK_FOR_INTERRUPTS();
+
+		if (!wasNonblocking)
+		{
+			PQsetnonblocking(pgConn, true);
+		}
+
+		/*
+		 * If busy, there might still be results already received and buffered
+		 * by the OS. As connection is in non-blocking mode, we can check for
+		 * that without blocking.
+		 */
+		if (PQisBusy(pgConn))
+		{
+			if (PQflush(pgConn) == -1)
+			{
+				/* write failed */
+				break;
+			}
+			if (PQconsumeInput(pgConn) == 0)
+			{
+				/* some low-level failure */
+				break;
+			}
+		}
+
+		/* clearing would require blocking IO, return */
+		if (PQisBusy(pgConn))
+		{
+			break;
+		}
+
+		result = PQgetResult(pgConn);
+		if (PQresultStatus(result) == PGRES_COPY_IN)
+		{
+			PQputCopyEnd(connection->pgConn, NULL);
```

Comment:
If we get here we're going to return false and close the connection, since we're almost always going to be busy in the next iteration. What if we just break here?

Minor note: When using cstore_fdw, `PQputCopyEnd` would trigger a "commit".